Source filtering

roberth opened the pull request:
Prevent rebuilds caused by changes in the parent directory, and/or sibling directories.

Reproduced in practice in `hercules-ci-agent` and easy to reproduce with a new test case.

srid commented:
@roberth I'm testing this on https://github.com/srid/haskell-multi-nix

Changing `README.md` and then running `nix run` rebuilds "foo" and "bar". Shouldn't this PR have fixed that?

roberth commented:
Looks ok when I run this:

```
[haskell-multi-nix]$ nix eval .#packages.x86_64-linux.bar.drvPath --override-input haskell-flake github:srid/haskell-flake/source-filter
warning: Git tree '/home/user/h/haskell-multi-nix' is dirty
warning: not writing modified lock file of flake 'git+file:///home/user/h/haskell-multi-nix':
• Updated input 'haskell-flake':
    'github:srid/haskell-flake/86695baa937a2aae6968c739a5493dee8ecde05c' (2023-02-19)
  → 'github:srid/haskell-flake/b05dc0395f69c9d2a560c791ab6c9f4676abff6e' (2023-03-11)
"/nix/store/0zc31pwd71pk7h46p9v2hj00hyl5vwdc-bar-0.1.0.0.drv"

[haskell-multi-nix]$ echo foo >>README.md 

[haskell-multi-nix]$ nix eval .#packages.x86_64-linux.bar.drvPath --override-input haskell-flake github:srid/haskell-flake/source-filter
warning: Git tree '/home/user/h/haskell-multi-nix' is dirty
warning: not writing modified lock file of flake 'git+file:///home/user/h/haskell-multi-nix':
• Updated input 'haskell-flake':
    'github:srid/haskell-flake/86695baa937a2aae6968c739a5493dee8ecde05c' (2023-02-19)
  → 'github:srid/haskell-flake/b05dc0395f69c9d2a560c791ab6c9f4676abff6e' (2023-03-11)
"/nix/store/0zc31pwd71pk7h46p9v2hj00hyl5vwdc-bar-0.1.0.0.drv"
```

Same derivation path, so that's a reused build.

roberth commented:
My tree was dirty before the experiment, but cleaning and redoing gives the same outcome.

srid commented:
I tried it on Linux, and the result is the same:

```
srid on pce haskell-multi-nix on  master via λ
❯ nix eval .#packages.x86_64-linux.bar.drvPath --override-input haskell-flake github:srid/haskell-flake/source-filter
warning: not writing modified lock file of flake 'git+file:///home/srid/code/haskell-multi-nix':
• Updated input 'haskell-flake':
    'github:srid/haskell-flake/56d8d9787ad8914bf54c2580a4dbb80adbe38f35' (2023-03-10)
  → 'github:srid/haskell-flake/b05dc0395f69c9d2a560c791ab6c9f4676abff6e' (2023-03-11)
"/nix/store/z4syg8rqq48ybiz10s3jn6ka60j88lh5-bar-0.1.0.0.drv"

srid on pce haskell-multi-nix on  master via λ took 11s
❯ echo foo >>README.md

srid on pce haskell-multi-nix on  master [!] via λ
❯ nix eval .#packages.x86_64-linux.bar.drvPath --override-input haskell-flake github:srid/haskell-flake/source-filter
warning: Git tree '/home/srid/code/haskell-multi-nix' is dirty
warning: not writing modified lock file of flake 'git+file:///home/srid/code/haskell-multi-nix':
• Updated input 'haskell-flake':
    'github:srid/haskell-flake/56d8d9787ad8914bf54c2580a4dbb80adbe38f35' (2023-03-10)
  → 'github:srid/haskell-flake/b05dc0395f69c9d2a560c791ab6c9f4676abff6e' (2023-03-11)
"/nix/store/ws0yd930drz9sm6if34v6lk15k36zd2w-bar-0.1.0.0.drv"

srid on pce haskell-multi-nix on  master [!] via λ took 3s
❯
```

Nix info,

```
❯ nix-info
system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.13.2, channels(root): "nixos-23.05pre449318.2caf4ef5005", nixpkgs: /nix/var/nix/profiles/per-user/root/channels/nixos
```

What could be different about your environment?

srid commented:
Strangely, it works for `./test/with-subdir` in this PR. Just not haskell-multi-nix ....

srid commented:
Oh, the cause was the use of *older* nixpkgs in haskell-multi-nix. Advancing it this way resolves it:

• Updated input 'nixpkgs':
    'github:nixos/nixpkgs/bbf77421ac51a7c93f5f0f760da99e4dbce614fa' (2022-11-09)
  → 'github:nixos/nixpkgs/bb31220cca6d044baa6dc2715b07497a2a7c4bc7' (2022-12-19)

Out of curiosity, what changed in nixpkgs recently (since 11-09) regarding this behaviour?

srid commented on nix/haskell-project.nix:
```diff
@@ -55,14 +58,31 @@ in
+            [
+              (pkgCfg: self.callCabal2nix name pkgCfg.root { })
+
+              # Avoid rebuilding because of changes in parent directories
+              (setSrc name pkgCfg.root)
```
What's the rationale behind the use of `lib.cleanSourceWith` somehow not triggering rebuilds that are otherwise triggered?

srid replied:
You may want to add a test to reproduce the rebuild such that only the use of `setSrc` can fix it.

srid replied:
> I removed the `srcSrc` call, but the tests still pass (on macOS at least)

Bizarrely, this is not the case with haskell-multi-nix. Your PR *does* prevent rebuild. Using `master` branch of haskell-flake causes rebuilds when README.md changes. What's different about `test/with-subdir` though?

srid replied:
Looking inside the two `.drv`, I see that the source hash changes, yet the tarball contents themselves are identical:

```
❯ nix run nixpkgs#pkgdiff /nix/store/9qy8hbsbawcw1crmr9nbljr3yyn5skzw-source.tar.gz /nix/store/f2daz0iv6dn8j6iw95s16rw5zn27ca8w-source.tar.gz
reading packages ...
WARNING: different package names ("9qy8hbsbawcw1crmr9nbljr3yyn5skzw" and "f2daz0iv6dn8j6iw95s16rw5zn27ca8w")
comparing packages ...
creating report ...
result: UNCHANGED
report: pkgdiff_reports/9qy8hbsbawcw1crmr9nbljr3yyn5skzw/source_to_source/changes_report.html
```

roberth replied:
> What's the rationale behind the use of `lib.cleanSourceWith` somehow not triggering rebuilds that are otherwise triggered?

`pkgCfg.root` is a value like `"${root}/mypackage"`, so something like `"/nix/store/...-source/mypackage"`, where `/nix/store/...-source` contains everything at the root of the flake.

`cleanSourceWith` does not copy things from the parent directory, so it returns a store path that only contains the relevant directory.

> Looking inside the two `.drv`, I see that the source hash changes, yet the tarball contents themselves are identical:

That means that `cabal sdist` is doing its job. You'll need the `src.src` - the source of the sdist build - to be identical too. Otherwise, the sdist builds are going to be different.

srid replied:
Thanks; added this in 600a179

srid reviewed (commented):
@roberth I removed the `srcSrc` call, but the tests still pass (on macOS at least). So the `setSrc`/`cleanSourceWith` part appears to be redundant (reducing this PR to mainly adding a test case to ensure that rebuilds don't happen).

```
❯ g diff
diff --git a/nix/haskell-project.nix b/nix/haskell-project.nix
index 0afc23d..033d5e5 100644
--- a/nix/haskell-project.nix
+++ b/nix/haskell-project.nix
@@ -76,9 +76,6 @@ in
             [
               (pkgCfg: self.callCabal2nix name pkgCfg.root { })
 
-              # Avoid rebuilding because of changes in parent directories
-              (setSrc name pkgCfg.root)
-
               # Make sure all files we use are included in the sdist, as a check
               # for release-worthiness.
               fromSdist

❯ ./runtest.sh 
[..]

||| All tests passed!
```

---

cf. https://github.com/srid/haskell-flake/issues/88

srid commented on test/with-subdir/flake.nix:
```diff
@@ -0,0 +1,35 @@
+    flake-parts.url = "github:hercules-ci/flake-parts/7c7a8bce3dffe71203dcd4276504d1cb49dfe05f";
+    check-flake.url = "github:srid/check-flake/48a17393ed4fcd523399d6602c283775b5127295";
+
+    haskell-multi-nix.url = "github:srid/haskell-multi-nix/7aed736571714ec12105ec110358998d70d59e34";
+    haskell-multi-nix.flake = false;
```
This is unused.

srid commented on test/with-subdir/flake.nix:
```diff
@@ -0,0 +1,35 @@
+  };
+  outputs = inputs@{ self, nixpkgs, flake-parts, ... }:
+    flake-parts.lib.mkFlake {
+      inputs = inputs // {
+        # work around https://github.com/NixOS/nix/pull/7796 for nix <2.14
+        self = self // { outPath = ./.; };
+      };
```
We won't need this workaround if you use `${NIX}` (instead of `nix`) in `runtest.sh`

srid commented:
> I removed the `srcSrc` call, but the tests still pass (on macOS at least).

Turns out this is because those `[[ .. == .. ]]` conditions always succeed on macOS's bash (3.2.57), but it is not a problem with bash from nixpkgs (5.2.15) - using which the tests fail as I expected.

Aside: the bash script is getting unwieldy; I'm starting to seriously think that we should write the tests is some other language.

srid commented:
I tested this PR in https://github.com/nammayatri/nammayatri/pull/109 and it does prevent rebuilds. Great!

srid reviewed (approved):
Gonna rebase and merge shortly.

roberth commented on nix/haskell-project.nix:
```diff
@@ -55,14 +58,38 @@ in
+          makeSrcAutonomous = name: root: pkg: hlib.overrideSrc
+            {
+              src =
+                # Since 'root' maybe a subdirectory of another source
+                # derivation, the purpose of cleanSourceWith here is to create a
+                # new (smaller) source derivation that is a copy of 'root' but
+                # ignores the parent source contents.
```
```suggestion
                # Since 'root' may be a subdirectory of a store path
                # (in string form, which means that it isn't automatically
                # copied), the purpose of cleanSourceWith here is to create a
                # new (smaller) store path that is a copy of 'root' but
                # does not contain the unrelated parent source contents.
```

It wasn't a derivation output at this point yet.

srid replied:
Thanks; done.
